### PR TITLE
Allow a user-provided function to preprocess the signature response

### PR DESCRIPTION
**Pull request by fd**
Example usage.

```js
var config = {
  // ...
  signResponseHandler: function(response) {
    return JSON.parse(response).signature;
  }
}
```

### CODE REVIEW

File: evaporate.js [hunk 1]
```diff
@@ -50,6 +50,7 @@
             signHeaders: {},
             awsLambda: null,
             awsLambdaFunction: null,
+            signResponseHandler: null,
```
Comment (bikeath1337) [inline evaporate.js]: We need to update the docs as well as part of this PR, please

File: evaporate.js [hunk 2]
```diff
@@ -1198,7 +1211,11 @@
                         authRequester.onFailedAuth(err);
                         return;
                     }
-                    authRequester.auth = JSON.parse(data.Payload);
+                    var payload = JSON.parse(data.Payload);
+                    if (typeof con.signResponseHandler === 'function') {
+                      payload = con.signResponseHandler(payload) || payload;
+                    }
```
Comment (bikeath1337) [inline evaporate.js]: Are you sure this duplication is correct? You've got essentially the same code as in lines 1152-1154.

Comment (fd) [inline evaporate.js]: These lines allow you to preprocess a response from AWS lambda.

File: evaporate.js (hunk 1 above)
Comment (fd) [inline evaporate.js]: I added documentation in 9e809bb

**Comment by fd** [comment]
@bikeath1337 We have been using this code for a while now. Are there any more things you would like me to change/fix before merging?

**Comment by bikeath1337** [comment]
Thanks, Simon. I added issue #148 for this. I also squashed and rebased your work on master, then added a commit that DRYs up references to the new signing method on my fork, branch issue_148. Could you take a look?

The only thing I would like to see are examples of the signing functions. Since the code touches the AWS Lambda path as well as the non-Lambda path, having examples of both use-cases would help developers help themselves.

Thanks again for your contribution.

**Comment by bikeath1337** [comment]
I think we're ready to consider this for inclusion, @fd but could you help me one more time to understand the use-case for this PR. I completely understand what we're doing, but I would like to be able to describe why this feature exists and when consumers of EvaporateJS would want to avail themselves of the feature. Also, what are the other ways to get this done, besides a JavaScript method, if any?

**Comment by fd** [comment]
@bikeath1337 My particular use-case it the following:
I have JSON API (generated from GRPC) and I want the signing requests to be handled like any other API request. In other words I want to be able to send back a JSON response (potentially with some extra information) when signing a AWS request. Like this my API remains free of weird edge cases where some requests have a special response formats (like the default `text/plain` response understood by EvaporateJS).

As a side-effect I can also send back detailed error messages in the case I want to reject an upload.

**Comment by bikeath1337** [comment]
Thanks so much for your contribution. The changes are now on master, ready for the next versioned release. https://github.com/TTLabs/EvaporateJS/commit/df27c50024cc693ec51b57c573119b1fb419edf3

**Comment by fd** [comment]
✨  thanks @bikeath1337 ✨

**Comment by bikeath1337** [comment]
Released as version 1.3.0.

**Pull request closed without merging**